added option to normalize histogram plot

all4one-max opened the pull request:
- Normalize button works well with X, but doesn't respond to ordinal values
- Normalize button does not work with stacked and layered histogram, necessary warnings have been added

fdebrain commented on app.py:
```diff
@@ -9,7 +9,6 @@
     plot_box,
     plot_donut,
     plot_histo,
-    plot_line,
```
@all4one-max Could you merge the main branch back to your branch in order to pull the latest code?

all4one-max replied:
Hey! I was trying to do that, but it didn't work out. so, I did a hard reset, saved my previous code, deleted my old branch, created the new branch and made the changes using the saved code.

btw for merge, I was using the following command.
```
git checkout my-branch
git merge main
```
this was not working.

fdebrain replied:
Can you share the error/reason those commands are not working?

all4one-max replied:
There wasn't any error, everything executed find but the code didn't merge.

fdebrain commented on app.py:
```diff
@@ -153,12 +151,18 @@ def generate_select_boxes(options_x, options_y, options_color, key_prefix):
             elif not col_y and not col_color:
-                st.subheader("Simple histogram")
-                plot_histo(df, col_x=col_x, bin=bins, ordinal=ordinal)
+                if ordinal and normalize:
+                    st.warning("Please select only one (Ordinal or Normalize)")
```
Ideally, we should aim to have both ordinal and normalize checkable at the same time. I can take a look at it, doesn't seem very simple at first glance.

all4one-max replied:
Hey! can you help me with this? I am not able to figure out how to do this

fdebrain replied:
Sure, I will take a closer look at it.

fdebrain commented on src/plots.py:
```diff
@@ -151,27 +152,43 @@ def plot_histo(
             "mark": {"type": "bar", **CONFIG_MARK, "blend": "normal"},
             "params": config_params,
+            "transform": [
+                {"bin": True, "field": col_x, "as": "bin_col_x", "maxbins": bin},
```
I would suggest the following to make normalized histogram responsible to the bin slider

```suggestion
                {"bin": config_bin, "field": col_x, "as": "bin_col_x"},
```

all4one-max replied:
Hey! made this change, and the slider is responsive with normalize checked now. That must mean binned and maxbins cannot be set simultaneously.

fdebrain commented on src/plots.py:
```diff
@@ -151,27 +152,43 @@ def plot_histo(
+                    "groupby": ["bin_col_x", "bin_col_x_end"],
+                },
+                {"joinaggregate": [{"op": "sum", "field": "xCount", "as": "TotalCount"}]},
+                {"calculate": "datum.xCount/datum.TotalCount", "as": "PercentOfTotal"},
```
I'm not sure why but when you add up the percentage in the histogram (eg for 2 bins), it doesn't add up to 100%.

all4one-max replied:
![image](https://user-images.githubusercontent.com/60205465/195492839-1a716c05-f0db-47ed-9c24-481a227f269c.png)
![image](https://user-images.githubusercontent.com/60205465/195492896-79f027fb-16e7-4bf8-98d5-834ab08455b9.png)

See in the 2 photos I check the percentage shown. I summed up the frequency of each bin and then checked the percentage of each bin, and it is correct. Moreover, if you add the percentage of all the bars it will sum to 100%.

fdebrain replied:
Trying with the titanic dataset and age variable, the sum is not equal to 100% due to the presence of NaNs. I'll add a filter to discard them when normalizing (without ordinal being checked).

fdebrain commented on src/plots.py:
```diff
@@ -151,27 +152,43 @@ def plot_histo(
                     "type": "ordinal" if ordinal else "quantitative",
                     "axis": {"labelAngle": -45 if ordinal else 0},
-                    "bin": config_bin,
+                    "bin": {"maxbins": bin, "binned": True} if normalize else config_bin,
```
Labels on the x-axis should ideally be responsive to the bin slider as well (like when normalize is unchecked)

all4one-max replied:
I think so this would be resolved by the change you mentioned above.

fdebrain commented:
@all4one-max Thanks for the work, there are a few bugs that I would like to fix before merging. Here is a summary:
- Make the normalized histogram responsive to the bins slider 
- X-axis labels & ticks should adapt to the bins slider as well
- Make it possible to have a normalized and ordinal histogram
- Fix the transform such that percentages add up to 100% (probably due to null values)

I can help with some of these if needed.

I'll open a separate issue to extend the normalization to stacked and layered histograms.

fdebrain commented on src/plots.py:
```diff
@@ -151,27 +152,43 @@ def plot_histo(
         data=df,
         spec={
             **CONFIG_MAIN,
             "mark": {"type": "bar", **CONFIG_MARK, "blend": "normal"},
```
This will make the plot look a bit nicer 

```suggestion
            "mark": {"type": "bar", **CONFIG_MARK, "blend": "normal", "binSpacing": 0},
```

all4one-max replied:
added to the code.

all4one-max committed:
add normalize option to histogram plots

all4one-max commented:
@fdebrain Hey! I tried managing the ordinal and normalize together but there's not much content available on the internet to work this out. I would request you merge with the current set of changes. If not possible then you may close the issue and assign me the heatmap one, I'll try that.

fdebrain commented:
> @fdebrain Hey! I tried managing the ordinal and normalize together but there's not much content available on the internet to work this out. I would request you merge with the current set of changes. If not possible then you may close the issue and assign me the heatmap one, I'll try that.

@all4one-max Thanks for your help! I'll give it a try and if I don't find a solution this weekend, I'll just merge and create a new issue.

fdebrain committed:
refactor(plot_histo)

A few improvements like:
- filter NaNs when normalizing 
- one line style when readable
- Remove unused argument col_y
- Field useless when counting

fdebrain committed:
refactor(app): histo widgets

Only show ordinal widget for simple & stacked/layered histo. 
Only show normalize widget for simple histo

fdebrain commented:
**What is done:**
- Implementation of basic normalization with responsive binning
- Ignoring NaNs when normalizing
- Warning messages for not yet implemented features regarding histogram normalization.

**What is to be done:** (will open a separate)
- X-axis labels & ticks should adapt to the bins slider as well
- Make it possible to have a normalized and ordinal histogram
- Extend normalization to stacked and layered histograms
